### PR TITLE
Pipenv 3.2.11 (new formula)

### PR DESCRIPTION
**Pull request by iraquitan**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
Add new homebrew formula for the pipenv project by @kennethreitz

### CODE REVIEW

File: Formula/pipenv.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+class Pipenv < Formula
+  include Language::Python::Virtualenv
+  desc "Sacred Marriage of Pipfile, Pip, & Virtualenv by @kennethreitz"
+  homepage "https://github.com/kennethreitz/pipenv"
+  url "https://github.com/kennethreitz/pipenv/archive/v3.2.10.tar.gz"
+  sha256 "c213e610406746c0a9e0c4f0060a859b1a5417afdafc905877b24539db7d0349"
+
+  depends_on :python if MacOS.version <= :lion
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Why `:lion`?

File: Formula/pipenv.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+class Pipenv < Formula
+  include Language::Python::Virtualenv
+  desc "Sacred Marriage of Pipfile, Pip, & Virtualenv by @kennethreitz"
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Don't need `kennethreitz` in here. A better description would be good, not clear what this means.

File: Formula/pipenv.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+class Pipenv < Formula
+  include Language::Python::Virtualenv
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Newline after this.

File: Formula/pipenv.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                              "--ignore-installed", buildpath
+    system libexec/"bin/pip", "uninstall", "-y", "pipenv"
+    venv.pip_install_and_link buildpath
+  end
+
+  def caveats; <<-EOS.undent
+      Proper casing should be used (e.g. Flask & Django).
+      Hashes are generated in lockfiles but not (yet) used for installation.
+      EOS
+  end
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Can remove these as they aren't Homebrew specific.

File: Formula/pipenv.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  test do
+    # Create a new project using Python 3
+    system "#{bin}/pipenv", "--three"
+    # Assert a Pipfile is created
+    assert_match "1\n", shell_output("ls | grep -ci Pipfile")
+    # Assert pipenv --help contains Commands
+    assert_match "Commands", shell_output("#{bin}/pipenv --help")
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Can remove this check.

File: Formula/pipenv.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def caveats; <<-EOS.undent
+      Proper casing should be used (e.g. Flask & Django).
+      Hashes are generated in lockfiles but not (yet) used for installation.
+      EOS
+  end
+
+  test do
+    # Create a new project using Python 3
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Can remove all these comments. They are self-explanatory from the code.

File: Formula/pipenv.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    system "#{bin}/pipenv", "--three"
+    # Assert a Pipfile is created
+    assert_match "1\n", shell_output("ls | grep -ci Pipfile")
+    # Assert pipenv --help contains Commands
+    assert_match "Commands", shell_output("#{bin}/pipenv --help")
+    # Install requests
+    system "#{bin}/pipenv", "install", "requests"
+    # Assert requests is now installed
+    assert_equal "1\n", shell_output("cat Pipfile | grep -ci requests")
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Use `assert_match` rather than `| grep`

File: Formula/pipenv.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    # Generate a Pipfile.lock file
+    system "#{bin}/pipenv", "lock"
+    # Assert Pipfile.lock is created
+    assert_equal "1\n", shell_output("ls | grep -ci Pipfile.lock")
```
Comment (MikeMcQuaid) [inline Formula/pipenv.rb]: Use `File.exist?` rather than `ls | grep`

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
Test is failing: https://bot.brew.sh/job/Homebrew%20Core/15574/version=sierra/testReport/junit/brew-test-bot/sierra/test_pipenv/

Happy to wave the newness audit on this.

File: Formula/pipenv.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
+class Pipenv < Formula
+  include Language::Python::Virtualenv
+
+  desc "Experimental virtualenv, and package manager for Python"
+  homepage "https://github.com/kennethreitz/pipenv"
+  url "https://github.com/kennethreitz/pipenv/archive/v3.2.10.tar.gz"
```
Comment (robsonpeixoto) [inline Formula/pipenv.rb]: Please, upgrade to 3.2.11. `sha256 "4dc7dcea983f0ecd8f3ca16e9249280271558bdffd0cc1bdca3bdf322f365d4e"`

**Comment by iraquitan** [comment]
@MikeMcQuaid Any help on why it is failing on testbot ?

**Comment by tdsmith** [comment]
Does it work for you locally? We're nesting virtualenvs here; I don't know how graceful that is.

File: Formula/pipenv.rb [hunk 10]
```diff
@@ -0,0 +1,131 @@
+class Pipenv < Formula
+  include Language::Python::Virtualenv
+
+  desc "Experimental virtualenv, and package manager for Python"
```
Comment (tdsmith) [inline Formula/pipenv.rb]: I'm still not sure what this means.

File: Formula/pipenv.rb [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      bash_completion.install "completions/pipenv"
+      fish_completion.install "completions/pipenv.fish"
+      zsh_completion.install "completions/_pipenv"
+    end
+  end
+
+  def caveats; <<-EOS.undent
+    To use Python 3 virtualenvs, `brew install python3`.
+    EOS
+  end
```
Comment (tdsmith) [inline Formula/pipenv.rb]: I'd get rid of this caveat as well.

File: Formula/pipenv.rb [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    sha256 "02f8102c2436bb03b3ee6dede1919d1dac8a427541652e5ec95171ec8adbc93a"
+  end
+
+  def install
+    virtualenv_install_with_resources
+    if build.with? "completions"
+      Dir.mkdir("completions")
+      system "_PIPENV_COMPLETE=source-bash #{bin}/pipenv > completions/pipenv"
+      system "_PIPENV_COMPLETE=source-fish #{bin}/pipenv > completions/pipenv.fish"
+      system "_PIPENV_COMPLETE=source-zsh #{bin}/pipenv > completions/_pipenv"
```
Comment (tdsmith) [inline Formula/pipenv.rb]: `bin` can have spaces in it; these should probably be quoted (or, better, don't invoke a shell for this).

File: Formula/pipenv.rb (hunk 10 above)
Comment (iraquitan) [inline Formula/pipenv.rb]: I'll describe better

**Comment by iraquitan** [comment]
@tdsmith Yes, it works locally. I performed the steps in testbot and worked. I'll try the new release that uses pew to manage virtualenvs.

**Comment by tdsmith** [comment]
Just checking -- when I say "locally" I mean running `brew test pipenv`. I can help troubleshoot later today if this is still snagging.

**Comment by nikolay** [comment]
Any luck?

**Comment by bfontaine** [comment]
Ping?

**Comment by JCount** [comment]
Closing per inactivity. Thank you for your work nonetheless @iraquitan. If you end up deciding to work on this again, feel free to open a new PR.

**Comment by tilgovi** [comment]
If I picked this up, would that be welcome?

**Comment by MikeMcQuaid** [comment]
@tilgovi If it would be useful to you: yes, definitely, thanks!

**Pull request closed without merging**